### PR TITLE
Remount support

### PR DESCRIPTION
**Pull request by anti-social**
Think there are many arguable places. Possibly you have better ideas.

### CODE REVIEW

**Comment by anti-social** [comment]
Some notes about parsing mountinfo file: there might be non-ascii characters.

Also in mount utility they only unescape octals: 
https://github.com/karelzak/util-linux/blob/master/libmount/src/tab_parse.c#L86
https://github.com/karelzak/util-linux/blob/master/lib/mangle.c#L51

File: src/mountinfo.rs [hunk 1]
```diff
@@ -0,0 +1,418 @@
...
+}
+
+fn parse_int(columns: &mut Iterator<Item=&[u8]>, row: &[u8], row_num: usize)
+    -> Result<c_ulong, ParseError>
+{
+    let col = try!(columns.next()
+        .ok_or_else(|| ParseError(
+            format!("Expected more fields"),
+            row_num, String::from_utf8_lossy(row).into_owned()))
+        .map(|v| String::from_utf8_lossy(v)));
```
Comment (tailhook) [inline src/mountinfo.rs]: It's unwise to use `String::from_utf8_lossy` here, it allocates a new string if value contains non-utf-8 bytes. But bytes that can be parsed are actually all-ascii. So you will allocate bytes and throw them away shortly

While it's not a performance issue in general case it also makes it hard to read.

File: src/bind.rs [hunk 2]
```diff
@@ -35,20 +36,29 @@ impl BindMount {
...
+    }
 
     /// Execute a bind mount
     pub fn bare_mount(self) -> Result<(), OSError> {
         let mut flags = flags::MS_BIND;
         if self.recursive {
             flags = flags | flags::MS_REC;
         }
+        if self.readonly {
+            flags = flags | flags::MS_RDONLY;
```
Comment (tailhook) [inline src/bind.rs]: This is useless. It doesn't work. You have to remount later to make a readonly bind mount.

File: src/remount.rs [hunk 3]
```diff
@@ -0,0 +1,439 @@
...
+fn apply_flag(flags: MsFlags, flag: MsFlags, set: Option<bool>) -> MsFlags {
+    match set {
+        Some(true) => flags | flag,
+        Some(false) => flags & !flag,
+        None => flags,
+    }
+}
+
+#[derive(Debug)]
+pub enum RemountError {
```
Comment (tailhook) [inline src/remount.rs]: Is there any reason you don't use `quick_error` here?

File: src/mountinfo.rs [hunk 4]
```diff
@@ -0,0 +1,418 @@
...
+    Ok(PathBuf::from(try!(parse_os_str(columns, row, row_num))))
+}
+
+fn unescape_octals(v: &mut Cow<[u8]>) {
+    let mut i = 0;
+    while i < v.len() {
+        if v[i] == b'\\' && v[i..].len() >= 4
+            && is_oct(v[i+1]) && is_oct(v[i+2]) && is_oct(v[i+3])
+        {
+            let t = v.to_mut().split_off(i);
```
Comment (tailhook) [inline src/mountinfo.rs]: Wow. You allocate a new vector here, for every single escape character.

I mean the complexity of this algo is `(number of characters in string) x (number of escapes)`

File: src/mountinfo.rs [hunk 5]
```diff
@@ -0,0 +1,418 @@
...
+use std::borrow::Cow;
+use std::error::Error;
+
+use libc::c_ulong;
+use libc::{MS_RDONLY, MS_NOSUID, MS_NODEV, MS_NOEXEC, MS_SYNCHRONOUS};
+use libc::{MS_MANDLOCK, MS_DIRSYNC, MS_NOATIME, MS_NODIRATIME};
+use libc::{MS_RELATIME, MS_STRICTATIME};
+
+#[derive(Debug)]
+pub struct ParseError(String, usize, String);
```
Comment (tailhook) [inline src/mountinfo.rs]: It's hard to reason what these fields mean. At least make it a struct with named fields. Better make a quick_error with contexts.

**Review by tailhook — COMMENTED** [review]
Well, it's all fine. But I don't like parsing very much. What I'd like to have is:

1. Decode utf-8 when whole `mountinfo` file is read
2. All string-like things in MountPoint should be slices in that string. Probably good idea is to make type along the lines of:

```
struct Slice<'a>(&'a str);

impl<'a> Slice<'a> {
  fn as_str(&self) -> Cow<'a, str>;
  fn as_os_str(&self) -> Cow<'a, OsStr>;
}
```
They should work as `String::from_utf8_lossy`, i.e. only allocate if at least one escape is in the string. (Bonus point: implement `PartialEq<&str> for Slice` directly but this might be overkill for the task).

For many fields `as_str` will never be called, for other there will almost never be any escaped characters.

We do many readonly bindmounts when running containers and we must reread whole mountinfo file in most cases, so it's important to have fast mountinfo.

Also decoding whole file at start will make your code simpler I think.

File: src/remount.rs [hunk 6]
```diff
@@ -0,0 +1,439 @@
...
+        Err(e) => Err(e),
+    }
+}
+
+fn get_mountpoint_flags_from(content: &[u8], path: &Path)
+    -> Result<Option<c_ulong>, RemountError>
+{
+    let mounts_parser = MountInfoParser::new(content);
+    let mut flags = None;
+    for mount_info_res in mounts_parser {
```
Comment (tailhook) [inline src/remount.rs]: Additional thought: You should use reverse iterator and break as soon as possible.

**Comment by tailhook** [comment]
By the way, have you verified that it solves https://github.com/tailhook/vagga/issues/358 ?

File: src/bind.rs (hunk 2 above)
Comment (anti-social) [reply src/bind.rs]: That's true. Didn't know that kernel just ignores `MS_RDONLY` flag for bind mount.

File: src/remount.rs (hunk 6 above)
Comment (anti-social) [reply src/remount.rs]: Had the same idea. So as I understand we need one more iterator: `MountInfoReverseParser`.

**Comment by anti-social** [comment]
I made tailhook/vagga#358 with additional mount call for read-only bind mounts so it worked for me.

**Comment by tailhook** [comment]
> I made tailhook/vagga#358 with additional mount call for read-only bind mounts so it worked for me.

I mean have you tried to remount on tmpfs or other differently configured volume?

Second point: the call should not be "additional" the flag in bind mount is just useless it should not be present there, it just makes code hard to read.

File: src/remount.rs (hunk 6 above)
Comment (tailhook) [reply src/remount.rs]: I think something like this should be sufficient:
```rust
for mount_info in content.split_lines().rev().map(parse_mountpoint) {
}
```

File: src/bind.rs [hunk 7]
```diff
@@ -66,10 +66,23 @@ impl BindMount {
...
         }
+        // do additional mount cause kernel just silently ignores MS_RDONLY
+        // flag for bind mount
+        if self.readonly {
+            let flags = flags | flags::MS_REMOUNT;
+            let rc = unsafe { mount(
+                    null(),
+                    self.target.as_ptr(),
+                    null(),
+                    flags.bits(),
```
Comment (tailhook) [inline src/bind.rs]: The ultimate point of this pull request is to use flags from the original mountpoint, right? And you don't use them here.

**Comment by anti-social** [comment]
> 1. Decode utf-8 when whole mountinfo file is read

There can be non-valid utf-8 bytes in the mountinfo file. And I think we should correctly process those files.

```
$ mkdir $'\377\377'
$ sudo mount -t tmpfs tmpfs $'\377\377'
$ python -c 'print(open(b"/proc/self/mountinfo", "rb").readlines()[-1])'
b'278 69 0:54 / /home/alexk/projects/libmount/\xff\xff rw,relatime shared:165 - tmpfs tmpfs rw\n'
```

2. I'll try to rid of needless memory allocations. In my first attempt borrow checker beat me.

**Comment by tailhook** [comment]
> There can be non-valid utf-8 bytes in the mountinfo file.

Okay, you're right. For some reason, I've thought kernel should escape all non-ascii characters, but it escapes control characters only :(

> In my first attempt borrow checker beat me.

Yeah, it can be hard, but I'm pretty sure it's doable. Don't hesitate to ask if you have specific troubles.

File: src/mountinfo.rs [hunk 8]
```diff
@@ -0,0 +1,462 @@
...
+    let bytes = try!(columns.next()
+        .ok_or_else(|| ParseError(
+            format!("Expected more fields"),
+            row_num, String::from_utf8_lossy(row).into_owned())));
+    let mut value = Cow::Borrowed(bytes);
+    unescape_octals(&mut value);
+    Ok(OsString::from_vec(value.into_owned()))
+}
+
+fn parse_int(columns: &mut Iterator<Item=&[u8]>, row: &[u8], row_num: usize)
```
Comment (tailhook) [inline src/mountinfo.rs]: You don't need your own int parser. Just use `std::str::from_utf8` and fail immediately when data is not utf-8.

**Comment by anti-social** [comment]
Have you any notes or suggestions?

**Comment by tailhook** [comment]
All was fine IIRC, but I have to do a closer look at error type(s).

File: src/remount.rs [hunk 9]
```diff
@@ -311,7 +311,7 @@ fn get_mountpoint_flags(path: &Path) -> Result<MsFlags, RemountError> {
         mpath.push(path);
         mpath
     };
-    let mut mountinfo_content = vec!();
+    let mut mountinfo_content = Vec::with_capacity(4 * 1024);
```
Comment (anti-social) [inline src/remount.rs:270]: Inside my containers `/proc/self/mountinfo` file has a size of about 3Kb.

**Comment by tailhook** [comment]
Okay, so error reporting story of this library is:

* we have a cheap `OSError` which returned from every operation here
* we have `Error` which provides nice descriptions and other things

I want `Remount` to stay with the same interface. It's important for forward compatibility. I mean if we will ever change parsing mechanisms of the library or even find a syscall to do the thing, or port the library to another system user don't have to adapt.

So what you need to do:
1. Create a **private** "mount info parsing error" with quick_error, with all options that can happen there, including IO and parser errors. It's better if parsing errors would be structured and not `format!`ted. There are only 3 of them, it's not hard to describe them in enum (say `RemountError`)
2. return `OSError(io::Error::new(io::ErrorKind::InvalidData,  RemountError::xxx()))`
3. check the output of explain, if that's not good enough, use `Error::is` / `Error::downcast_ref` to find out that the underlying error is `RemountError` and make this specific case look nicer

**Comment by tailhook** [comment]
Okay, looks good. I've squashed and merged this in.

Also this lacks `.readonly()` method in `Bind`, you may submit that as a separate PR.

**Pull request closed without merging**